File: packages/javascript/lib/format/fileSizeFormat.ts
import type { FileSizeOptions, PositiveNumber } from '../_types/global.js';
import { fileSizeParts, resolveUnitTable, unitLabel } from './fileSizeParts.js';

export function fileSizeFormat<N extends number>(
	bytes: PositiveNumber<N>,
	decimals = 2,
	ceil = false,
	options?: FileSizeOptions
): string {
	const { standard, unitDisplay } = { ...options };
	const { value, exponent } = fileSizeParts(bytes, options);

	const rounded = ceil ? Math.ceil(value) : parseFloat(value.toFixed(decimals < 0 ? 0 : decimals));

	// The label is taken from the rounded number, so a value that rounds to one is not
	// written as `1 Megabytes`.
	return `${rounded} ${unitLabel(resolveUnitTable(standard), exponent, unitDisplay, rounded)}`;
}

File: packages/python/qsu/format/fileSizeFormat.py
import math
from typing import Optional

from .fileSizeParts import _unitLabel, _unitTable, fileSizeParts


def _numStr(value: float) -> str:
	if isinstance(value, float) and value.is_integer():
		return str(int(value))
	return str(value)


def _toFixed(value: float, digits: int) -> float:
	# Mirror JS Number.prototype.toFixed followed by parseFloat (drops trailing zeros).
	formatted = f'{value:.{digits}f}'
	return float(formatted)


def fileSizeFormat(
	bytes: float,
	decimals: int = 2,
	ceil: bool = False,
	standard: Optional[str] = None,
	unitDisplay: Optional[str] = None,
) -> str:
	parts = fileSizeParts(bytes, standard, unitDisplay)

	if ceil:
		value = math.ceil(parts['value'])
	else:
		value = _toFixed(parts['value'], 0 if decimals < 0 else decimals)

	# The label is taken from the rounded number, so a value that rounds to one is not
	# written as `1 Megabytes`.
	label = _unitLabel(_unitTable(standard), parts['exponent'], unitDisplay, value)

	return f'{_numStr(value)} {label}'
